processing/src/main/java/io/druid/query/aggregation/cardinality/CardinalityAggregatorFactory.java

```diff
@@ -94,11 +103,102 @@ public static Object estimateCardinality(Object object)
+      switch(type) {
+        case STRING:
+          return new StringCardinalityAggregatorTypeHelper();
+        default:
+          return null;
+      }
+    }
+  }
+
+  public interface CardinalityAggregatorTypeHelper<ValueSelectorType extends ColumnValueSelector> extends QueryTypeHelper
+  {
+    /**
+     * Used by CardinalityAggregator.
+     *
+     * Retrieve the current row from dimSelector and add the row values to the hasher.
+     *
+     * @param dimSelector Dimension value selector
+     * @param hasher Hasher used for cardinality aggregator calculations
+     */
+    void hashRow(ValueSelectorType dimSelector, Hasher hasher);
+
+
+    /**
+     * Used by CardinalityAggregator.
+     *
+     * Retrieve the current row from dimSelector and add the row values to the hasher.
+     * @param dimSelector Dimension value selector
+     * @param collector HLL collector used for cardinality aggregator calculations
+     */
+    void hashValues(ValueSelectorType dimSelector, HyperLogLogCollector collector);
+  }
+
+  public static class StringCardinalityAggregatorTypeHelper implements CardinalityAggregatorTypeHelper<DimensionSelector>
+  {
+    public static final String CARDINALITY_AGG_NULL_STRING = "\u0000";
+    public static final char CARDINALITY_AGG_SEPARATOR = '\u0001';
+
+    @Override
+    public void hashRow(DimensionSelector dimSelector, Hasher hasher)
+    {
+      final IndexedInts row = dimSelector.getRow();
+      final int size = row.size();
+      // nothing to add to hasher if size == 0, only handle size == 1 and size != 0 cases.
+      if (size == 1) {
+        final String value = dimSelector.lookupName(row.get(0));
+        hasher.putUnencodedChars(convertValueForCardinalityAggregator(value));
+      } else if (size != 0) {
+        final String[] values = new String[size];
+        for (int i = 0; i < size; ++i) {
+          final String value = dimSelector.lookupName(row.get(i));
+          values[i] = convertValueForCardinalityAggregator(value);
+        }
+        // Values need to be sorted to ensure consistent multi-value ordering across different segments
+        Arrays.sort(values);
+        for (int i = 0; i < size; ++i) {
+          if (i != 0) {
+            hasher.putChar(CARDINALITY_AGG_SEPARATOR);
+          }
+          hasher.putUnencodedChars(values[i]);
+        }
+      }
+    }
+
+    @Override
+    public void hashValues(DimensionSelector dimSelector, HyperLogLogCollector collector)
+    {
+      for (IntIterator rowIt = dimSelector.getRow().iterator(); rowIt.hasNext(); ) {
+        int index = rowIt.nextInt();
+        final String value = dimSelector.lookupName(index);
+        collector.add(CardinalityAggregator.hashFn.hashUnencodedChars(convertValueForCardinalityAggregator(value)).asBytes());
+      }
+    }
+
+    // CardinalityAggregator has a special representation for nulls
+    private String convertValueForCardinalityAggregator(String value)
```

Comment:
Maybe rename it to something like `nullToSpecial()` and remove the comment above? Like `Strings.nullToEmpty()`.